packages/flutter/lib/src/material/focus_highlight.dart

```diff
@@ -0,0 +1,212 @@
+}
+
+class _FocusHighlightState extends State<FocusHighlight> {
+  bool _hasFocus;
+  FocusNode _node;
+
+  @override
+  void didUpdateWidget(FocusHighlight oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    final FocusNode newNode = Focus.of(context);
+    if (newNode != _node) {
+      _node.removeListener(_handleFocusChanged);
+      _node = newNode;
+      _node.addListener(_handleFocusChanged);
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _node = Focus.of(context);
+    _node.addListener(_handleFocusChanged);
+    _hasFocus = _node.hasFocus;
+  }
+
+  @override
+  void dispose() {
+    _node?.removeListener(_handleFocusChanged);
+    super.dispose();
+  }
+
+  void _handleFocusChanged() {
+    if (_hasFocus != _node.hasFocus) {
+      setState(() {
+        _hasFocus = _node.hasFocus;
+      });
+    }
+  }
+
+  Decoration _getDecoration(BuildContext context) {
+    if (_hasFocus) {
+      return widget.focusedDecoration ?? Theme.of(context).focusHighlightTheme.focusedDecoration;
+    } else {
+      return widget.unfocusedDecoration ?? Theme.of(context).focusHighlightTheme.unfocusedDecoration;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedContainer(
+      padding: EdgeInsets.zero,
+      decoration: _getDecoration(context),
+      duration: widget.duration ?? Theme.of(context).focusHighlightTheme.focusAnimationDuration,
+      curve: widget.curve ?? Theme.of(context).focusHighlightTheme.focusAnimationCurve,
+      child: widget.child,
+    );
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<Decoration>('decoration', widget.focusedDecoration, defaultValue: null));
+    properties.add(DiagnosticsProperty<Decoration>('unfocusedDecoration', widget.unfocusedDecoration, defaultValue: null));
+    properties.add(DiagnosticsProperty<Duration>('duration', widget.duration, defaultValue: null));
+    properties.add(DiagnosticsProperty<Curve>('curve', widget.curve, defaultValue: null));
+  }
+}
+
+/// Defines the appearance of [FocusHighlight]s.
+///
+/// This class is used to define the value of [ThemeData.focusHighlightTheme]. The
+/// [FocusHighlight] widget uses the current focusable theme to initialize
+/// [FocusHighlight] properties (like [FocusHighlight.focusedDecoration]) when
+/// they are not supplied.
+class FocusHighlightThemeData extends Diagnosticable {
+  /// Creates a const FocusThemeData object.
+  ///
+  /// All of the arguments are optional, and will use fallback values if not
+  /// specified.
+  const FocusHighlightThemeData({
+    Decoration focusedDecoration,
+    Decoration unfocusedDecoration,
+    Curve focusAnimationCurve,
+    Duration focusAnimationDuration,
+  })  : focusedDecoration = focusedDecoration ?? _defaultFocusedDecoration,
```

Comment:
these are const, so i think you can make them actually be defaults and assert non-null?

Reply:
Yes, I can, but I did it that way on purpose to avoid needing to know about the defaults if someone subclasses it, and passes its arguments to `super`: they would need to have their own defaults for things that weren't specified.  The way I have it, they could just be null, and they'd get the base class default. Let me know if you prefer the way you suggested, though, and I'll change it.

Reply:
we do it the other way in most places. it has the advantage of being more self-documenting.

if we feel like this is too much of a burden on subclasses, we should push back on dart.

Reply:
OK, I'll switch to defaults.  I do think this is a common source of errors in Dart.  Non-nullable types would go a long way towards solving it, though.